src/pairwise.jl

```diff
@@ -122,6 +122,33 @@ function _pairwise!(f, dest::AbstractMatrix, x, y;
     return _pairwise!(Val(skipmissing), f, dest, x′, y′, symmetric)
 end
 
+if VERSION >= v"1.6.0-DEV"
+    # Function has moved in Julia 1.7
+    if isdefined(Base, :typejoin_union_tuple)
+        using Base: typejoin_union_tuple
+    else
+        using Base.Broadcast: typejoin_union_tuple
+    end
+else
+    typejoin_union_tuple(::Type) = Any
+end
+
+# Identical to `Base.promote_typejoin` except that it uses `promote_type`
+# instead of `typejoin` to combine members of `Union` types
+function promote_type_union(::Type{T}) where T
+    if T === Union{}
+        return Union{}
+    elseif T isa UnionAll
+        return Any # TODO: compute more precise bounds
+    elseif T isa Union
+        return promote_type(promote_type_union(T.a), promote_type_union(T.b))
+    elseif T <: Tuple
+        return typejoin_union_tuple(T)
+    else
+        return T
```

Comment:
Is line 148 covered in tests?

Reply:
It's covered as `promote_type_union` calls itself recursively, but I've added two tests for clarity.